control-plane/connect-inject/namespace/namespace_controller_ent_test.go

```diff
@@ -0,0 +1,413 @@
+			EnableNSMirroring:          tc.enableNSMirroring,
+			NSMirroringPrefix:          tc.nsMirrorPrefix,
+			ConsulDestinationNamespace: tc.destinationNamespace,
+		}
+
+		if tc.kubeNamespaceName == "" {
+			tc.kubeNamespaceName = testNamespaceName
+		}
+
+		namespacedName := types.NamespacedName{
+			Name: tc.kubeNamespaceName,
+		}
+
+		resp, err := nc.Reconcile(context.Background(), ctrl.Request{
+			NamespacedName: namespacedName,
+		})
+		require.NoError(t, err)
+		require.False(t, resp.Requeue)
+
+		if tc.existingConsulNamespace != nil {
+			expectedNamespaceMatches(t, testClient.APIClient, tc.existingConsulNamespace.Name, tc.partition, tc.expectedConsulNamespace)
+		} else {
+			expectedNamespaceMatches(t, testClient.APIClient, testNamespaceName, tc.partition, tc.expectedConsulNamespace)
+		}
+	}
+
+	testCases := []testCase{
+		{
+			name:                    "destination namespace with default is not cleaned up",
+			existingConsulNamespace: getNamespace(constants.DefaultConsulNS, "", false),
+			expectedConsulNamespace: getNamespace(constants.DefaultConsulNS, "", false),
+		},
+		{
+			name:                    "destination namespace with non-default is not cleaned up",
+			destinationNamespace:    "bar",
+			existingConsulNamespace: getNamespace("bar", "", false),
+			expectedConsulNamespace: getNamespace("bar", "", false),
+		},
+		{
+			name:                    "destination namespace with non-default is not cleaned up, non-default partition",
+			destinationNamespace:    "bar",
+			partition:               "baz",
+			existingConsulNamespace: getNamespace("bar", "baz", false),
+			expectedConsulNamespace: getNamespace("bar", "baz", false),
+		},
+		{
+			name:                    "mirrored namespaces",
+			enableNSMirroring:       true,
+			existingConsulNamespace: getNamespace(testNamespaceName, "", false),
+		},
+		{
+			name:                    "mirrored namespaces but it's the default namespace",
+			kubeNamespaceName:       metav1.NamespaceDefault,
+			enableNSMirroring:       true,
+			existingConsulNamespace: getNamespace(constants.DefaultConsulNS, "", false),
+			expectedConsulNamespace: getNamespace(constants.DefaultConsulNS, "", false), // Don't ever delete the Consul default NS
+		},
+		{
+			name:                    "mirrored namespaces, non-default partition",
+			partition:               "baz",
+			enableNSMirroring:       true,
+			existingConsulNamespace: getNamespace(testNamespaceName, "baz", false),
+		},
+		{
+			name:                    "mirrored namespaces with prefix",
+			nsMirrorPrefix:          "k8s-",
+			enableNSMirroring:       true,
+			existingConsulNamespace: getNamespace("k8s-foo", "", false),
+		},
+		{
+			name:                    "mirrored namespaces with prefix, non-default partition",
+			partition:               "baz",
+			nsMirrorPrefix:          "k8s-",
+			enableNSMirroring:       true,
+			existingConsulNamespace: getNamespace("k8s-foo", "", false),
```

Comment:
I think this one needs a partition in existing?

```suggestion
			existingConsulNamespace: getNamespace("k8s-foo", "bar", false),
```

Reply:
Yeah, it should be `baz`. I guess it succeeds since the delete works even if the resource doesn't exist (some kind of idempotency?).

Reply:
Lol oops.

But yeah, w/ idempotent deletes and lacking some feedback that it caused a change to occur, it's a bit tricky to rule out false negatives. Could do a "list all namespaces -> count" check in the run function to rule those out?